Var dicts

nguy opened the pull request:
This PR is not complete yet. I want to get feedback. When changing the readers to dictionaries, it is required to modify the ```DropSizeDistribution``` object class. 

@josephhardinee can you:
1. Verify the changes submitted work, i.e. calculations perform as expected
2. Should we save spread, bin_edges, diameter as dictionaries or just arrays?

All **aux** readers will need updated.

nguy commented:
ping @josephhardinee Let me know when you get to this so I can adjust code accordingly and update other needed areas.

josephhardinee commented on pydisdrometer/io/JWDReader.py:
```diff
@@ -93,8 +95,37 @@ def _read_file(self):
-        self.rain_rate = np.ma.array(self.rain_rate)
+
+        self.fields['Nd'] = common.var_to_dict(
+            'Nd', np.ma.array(self.Nd), 'm^-3',
```
The units should be 'm^-3mm^-1.

nguy replied:
Oops, nice catch. Fixed.

josephhardinee commented:
1) It all appears to work on mine. 
2) I think probably as dicts to remain consistent. 

I'll add comments to the diffset. I saw a few minor things but all in all it appears to be pretty good!

josephhardinee commented on pydisdrometer/io/ParsivelNasaGVReader.py:
```diff
@@ -98,22 +98,31 @@ def __init__(self, filename, campaign, skip_header):
+    def _prep_data(self):
+        self.fields = {}
+        self.fields['Nd'] = common.var_to_dict(
+            'Nd', np.ma.array(self.Nd), 'm^-3',
```
Same with the units here(add the per millimeter)

nguy replied:
Done.

josephhardinee commented on pydisdrometer/io/ParsivelReader.py:
```diff
@@ -17,10 +20,11 @@ def read_parsivel(filename):
-                               rain_rate=reader.rain_rate, velocity=reader.velocity,
-                               Z=reader.Z, num_particles=reader.num_particles,
-                               bin_edges=reader.bin_edges, diameter=reader.diameter)
+##    dsd = DropSizeDistribution(reader.time, reader.Nd, reader.spread,
```
You can go ahead and delete the commented out lines.

nguy commented:
Are spread, bin_edges, diameter in millimeters?

josephhardinee commented:
Yes, they should be. A few file formats have them in something else and I try to make sure to always convert them to mm. Just makes life easier.

nguy commented:
Cool, I wanted to verify. So I'll start making more changes and push them up here for you to review also before merger.

nguy commented:
@josephhardinee I added dictionary for spread, diameter, bin_edges. Converted the entire "io" code base. Can you check again real fast to see if I didn't break anything?

nguy commented:
Most of the readers in the the ```aux_readers``` seciton need work on the time variable to calculate time. Specifically need to grab the date information of the file.

josephhardinee commented on pydisdrometer/io/JWDReader.py:
```diff
@@ -1,7 +1,8 @@
 # -*- coding: utf-8 -*-
 import numpy as np
+from netCDF4 import num2date, date2num
+
```
This needs a from .io import common

josephhardinee commented on pydisdrometer/aux_readers/GPMApuWallopsRawReader.py:
```diff
@@ -1,11 +1,7 @@
 import numpy as np
-import numpy.ma as ma
 from ..DropSizeDistribution import DropSizeDistribution
 
```
This needs a  from ..io import common

josephhardinee replied:
This needs a from ..io import common

josephhardinee commented on pydisdrometer/io/ParsivelNasaGVReader.py:
```diff
@@ -98,22 +92,34 @@ def __init__(self, filename, campaign, skip_header):
+        self.bin_edges = common.var_to_dict(
+            'bin_edges',
+            np.hstack(
+                (0, self.diameter + np.array(self.spread) / 2)),
```
I think this should be self.diameter['data']

nguy commented:
I think I addressed all of these. Good catches!

nguy commented:
Should we merge this or fix the time issues in the aux readers?
